src/python/pants/rules/core/lint.py

```diff
@@ -56,6 +56,29 @@ class LintOptions(GoalSubsystem):
 
     required_union_implementations = (Linter,)
 
+    @classmethod
+    def register_options(cls, register) -> None:
+        super().register_options(register)
+        register(
+            "--per-target-caching",
+            advanced=True,
+            type=bool,
+            default=False,
+            help=(
+                "Rather than running all targets in a single batch, run each target as a "
+                "separate process. Why do this? You'll get many more cache hits. Additionally, for "
+                "Python users, if you have some targets that only work with Python 2 and some that "
```

Comment:
I don't love the mention of Python in this generic location, but I guess it's fine. 

Can we not finesse the interpreter issue in the implementation? What happens in the py2/py3 case today with --no-per-target-caching ?

Reply:
Hm..we delegate everything to Pex and I think we are passing the arguments incorrectly to Pex. Right now, if you have a target that's Py2 only and a target that's Py3 only, we pass to Pex `['--interpreter-constraint', 'CPython==2.7.*', '--interpreter-constraint', 'CPython>=3.6.*']`. Multiple `--interpreter-constraints` to Pex OR the constraints, when I think we should be ANDing..

I'll open a ticket to investigate and ensure V2 properly supports interpreter constraints when there are multiple targets. Sounds like many users still have Python 2.

Reply:
Well, anding those constraints would create an unsatisfiable constraint no? 

I guess I meant could we partition the work by interpreter constraint, and do one batch per constraint?

Reply:
> Well, anding those constraints would create an unsatisfiable constraint no?

Yes, it would, just as V1 does. Erroring is an improvement on silently doing the wrong thing.

> I guess I meant could we partition the work by interpreter constraint, and do one batch per constraint?

Yes, we can, although that will require more work to do and is unclear how we would achieve it at the moment. Currently, we expect one set of targets -> one single LintResult. Here, we'd be saying that sometimes one set of targets -> multiple LintResults. That violates the assumption of `await Get[LintResult(Linter, linter(targets))`.

I'm sure it's possible, but not clear to me at all how to do it.